File: android/src/main/kotlin/dev/taskflow/TaskFlowPlugin.kt
package dev.taskflow

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** TaskFlowPlugin */
class TaskFlowPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var context: Context
  private var eventSink: EventChannel.EventSink? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
    channel = MethodChannel(binding.binaryMessenger, "dev.taskflow/channel")
    eventChannel = EventChannel(binding.binaryMessenger, "dev.taskflow/events")
    channel.setMethodCallHandler(this)
    eventChannel.setStreamHandler(
      object : EventChannel.StreamHandler {
        override fun onListen(args: Any?, sink: EventChannel.EventSink?) {
          eventSink = sink
        }

        override fun onCancel(args: Any?) {
          eventSink = null
        }
      }
    )
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "initialize" -> handleInitialize(result)
      "enqueue" -> handleEnqueue(call, result)
      "enqueueChain" -> handleEnqueueChain(call, result)
      "schedule" -> handleSchedule(call, result)
      "reschedule" -> handleReschedule(call, result)
      "unschedule" -> handleUnschedule(call, result)
      "cancel" -> handleCancel(call, result)
      "cancelExecution" -> handleCancelExecution(call, result)
      "cancelChain" -> handleCancelChain(call, result)
      "cancelByTag" -> handleCancelByTag(call, result)
      "cancelAll" -> handleCancelAll(result)
      "getStatus" -> handleGetStatus(call, result)
      "getAllTasks" -> handleGetAllTasks(result)
      "getTasksByTag" -> handleGetTasksByTag(call, result)
      "reportProgress" -> handleReportProgress(call, result)
      else -> result.notImplemented()
    }
  }

  private fun handleInitialize(result: Result) {
    // Initialize WorkManager and other platform components
    result.success(null)
  }

  private fun handleEnqueue(call: MethodCall, result: Result) {
    // TODO: Implement task enqueuing via WorkManager
    result.success("dummy-exec-id")
  }

  private fun handleEnqueueChain(call: MethodCall, result: Result) {
    // TODO: Implement chain enqueuing
    result.success("dummy-chain-id")
  }

  private fun handleSchedule(call: MethodCall, result: Result) {
    // TODO: Implement periodic scheduling
    result.success(null)
  }

  private fun handleReschedule(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleUnschedule(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleCancel(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleCancelExecution(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleCancelChain(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleCancelByTag(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleCancelAll(result: Result) {
    result.success(null)
  }

  private fun handleGetStatus(call: MethodCall, result: Result) {
    result.success(null)
  }

  private fun handleGetAllTasks(result: Result) {
    result.success(emptyList<Any>())
  }

  private fun handleGetTasksByTag(call: MethodCall, result: Result) {
    result.success(emptyList<Any>())
  }

  private fun handleReportProgress(call: MethodCall, result: Result) {
    result.success(null)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }
}

File: ios/Classes/TaskflowPlugin.swift
import BackgroundTasks
import Flutter
import UIKit

public class TaskFlowPlugin: NSObject, FlutterPlugin {
  private static var methodChannel: FlutterMethodChannel?
  private static var eventChannel: FlutterEventChannel?
  private static var eventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    methodChannel = FlutterMethodChannel(
      name: "dev.taskflow/channel",
      binaryMessenger: messenger
    )
    eventChannel = FlutterEventChannel(
      name: "dev.taskflow/events",
      binaryMessenger: messenger
    )

    let instance = TaskFlowPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel!)
    eventChannel?.setStreamHandler(instance)

    // Register BGTask identifiers early
    TaskScheduler.shared.registerAllTasks()
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      handleInitialize(result)
    case "enqueue":
      handleEnqueue(call, result)
    case "enqueueChain":
      handleEnqueueChain(call, result)
    case "schedule":
      handleSchedule(call, result)
    case "reschedule":
      handleReschedule(call, result)
    case "unschedule":
      handleUnschedule(call, result)
    case "cancel":
      handleCancel(call, result)
    case "cancelExecution":
      handleCancelExecution(call, result)
    case "cancelChain":
      handleCancelChain(call, result)
    case "cancelByTag":
      handleCancelByTag(call, result)
    case "cancelAll":
      handleCancelAll(result)
    case "getStatus":
      handleGetStatus(call, result)
    case "getAllTasks":
      handleGetAllTasks(result)
    case "getTasksByTag":
      handleGetTasksByTag(call, result)
    case "reportProgress":
      handleReportProgress(call, result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleInitialize(_ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleEnqueue(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let executionId = "exec-\(Date().timeIntervalSince1970)-\(Int.random(in: 0...999))"
    result(executionId)

    // Simulate task execution with status updates
    simulateTaskExecution(executionId)
  }

  private func simulateTaskExecution(_ executionId: String) {
    DispatchQueue.global().async {
      // Emit Queued status
      DispatchQueue.main.async {
        TaskFlowPlugin.eventSink?([
          "executionId": executionId,
          "taskName": "exampleTask",
          "type": "queued"
        ])
      }

      Thread.sleep(forTimeInterval: 0.5)

      // Emit Running status with progress
      for i in 1...5 {
        Thread.sleep(forTimeInterval: 0.4)
        DispatchQueue.main.async {
          TaskFlowPlugin.eventSink?([
            "executionId": executionId,
            "taskName": "exampleTask",
            "type": "running",
            "progress": Double(i) * 0.2
          ])
        }
      }

      Thread.sleep(forTimeInterval: 0.5)

      // Emit Succeeded status
      DispatchQueue.main.async {
        TaskFlowPlugin.eventSink?([
          "executionId": executionId,
          "taskName": "exampleTask",
          "type": "succeeded",
          "data": ["result": "Task completed successfully!"]
        ])
      }
    }
  }

  private func handleEnqueueChain(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result("dummy-chain-id")
  }

  private func handleSchedule(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleReschedule(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleUnschedule(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleCancel(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleCancelExecution(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleCancelChain(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleCancelByTag(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleCancelAll(_ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleGetStatus(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }

  private func handleGetAllTasks(_ result: @escaping FlutterResult) {
    result([])
  }

  private func handleGetTasksByTag(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result([])
  }

  private func handleReportProgress(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(nil)
  }
}

extension TaskFlowPlugin: FlutterStreamHandler {
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    TaskFlowPlugin.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    TaskFlowPlugin.eventSink = nil
    return nil
  }
}
